### ISSUE TITLE
Mactype not work with firefox 69.0(64 bit)

### ISSUE THREAD

**Issue by violenceandroid** [snowie2000/mactype#604]

**Comment by JB330** [comment]
Yes it does not work with FF69. So I had to downgrade FF68.0.2. Please help us snowie2000 T.T

**Comment by snowie2000** [comment]
Please enable directwrite option.

**Comment by JB330** [comment]
> 
> 
> Please enable directwrite option.

In .ini ? or firefox? 
I've already added in .ini directwrite=1 but it still not works.

**Comment by snowie2000** [comment]
In the ini. 
What’s the status of Firefox in the process manager ?

**Comment by maidonghu** [comment]
> In the ini.
> What’s the status of Firefox in the process manager ?

Running
But actually, it does not work.

**Comment by snowie2000** [comment]
If you have disabled directwrite before,  you have to re-enable it.

**Comment by neavo** [comment]
我也有这个问题 。。。
gfx.direct2d.disabled = true，关闭设置里的“启用硬件加速”，然后在配置文件ini里面设置了DirectWrite=1，在托盘图标的详细信息里也可以看到“firefox.exe”已启用了，但是实际上依然是无效的
折腾了一晚上，重启了好多次电脑，应该不是需要re-enable的问题，我用的配置文件是自带的FT Opt

**Comment by snowie2000** [comment]
@neavo 你都关闭了firefox的directwrite功能了，mactype又怎么会生效呢？

以前能生效是因为cairo是基于gdi的，而现在cairo已经移除了，你又关闭了dw当然不能用了

**Comment by neavo** [comment]
![1](https://user-images.githubusercontent.com/998676/64486904-e65e9980-d265-11e9-8ebe-f3aee2b2e855.jpg)
![2](https://user-images.githubusercontent.com/998676/64486905-e9598a00-d265-11e9-9d95-009e3258a7a5.jpg)

在FF69里，不论“gfx.direct2d.disabled”、“启用硬件加速”这几个组合如何排列组合，显示效果始终是上图，而在FF68ESR上，关闭direct2d并调整cario的顺序到第一位后，显示效果是下图这个样子的。
我理解的没错的话，上图应该是渲染没有成功生效吧？

**Comment by snowie2000** [comment]
正常，这就是directwrite渲染的效果，dw的效果和gdi是不同的，你可以和关闭mactype做对比

**Comment by neavo** [comment]
唔，晚上又测试了一下，好像开关MacType的显示效果并没有什么区别 。。。。有点纳闷

**Comment by JB330** [comment]
Tested newly downloaded fresh FF69, completely default setting, with default Mactype ini, 
It doesn't work. Can't see any difference between on and off.

**Comment by snowie2000** [comment]
If you have previously replaced Microsoft yahei with unhinted version, there should be no significant difference.

**Comment by SolidSapphire** [comment]
Is there a way to add back Cairo to firefox? also can you please do a tutorial on how to get this working for Firefox as a lot of users are confused.

**Comment by jmbattle0** [comment]
It's working here on my Win7 box.

I uninstalled the previous public release of MacType (pretty old), then installed the latest MacType beta.  Then I reset the Firefox profile and everything is working fine.  I'm currently using the default MacType .ini, but may tweak it later.

**Comment by snowie2000** [comment]
> Is there a way to add back Cairo to firefox? also can you please do a tutorial on how to get this working for Firefox as a lot of users are confused.

No, as stated in its bug tracker: https://hg.mozilla.org/mozilla-central/rev/46aad88d3b36 , https://bugzilla.mozilla.org/show_bug.cgi?id=1559810 cairo is blocked or removed. You have no way to add it back (unless you would like to build your firefox from source)

**Comment by SolidSapphire** [comment]
So what is the official status? Is it compatible? I'm currently using  Xsung Blade.

**Comment by snowie2000** [comment]
It's compatible with DirectWrite support.
It's just not as good as it was with cairo.

**Comment by JB330** [comment]
I've been tried with OTF font in FF69. So I Changed to TTF font and finally It works. But rendering quality is not good as cairo. Back to the FF68..

**Comment by PiotrGrochowski** [comment]
It is only possible to use GDI rendering in Firefox ESR 52.4.1 (32-bit).

The problem with DirectWrite is that the text is rendered with way too many dark gray pixels, causing text to become unrealistically thicker and more distorted. On the other hand, GDI uses correct gamma and fully respects font hinting, meaning text renders exactly as the type designer intended.
